cellbender/remove_background/data/dataset.py

```diff
@@ -1062,6 +1078,105 @@ def get_matrix_from_cellranger_h5(filename: str) \
+            in the list is the number of genomes in the dataset.  Each numpy
+            array contains the string names of genes in the genome, which
+            correspond to the columns in the out['matrix'].
+        out['gene_ids']: List of numpy arrays, where the number of elements
+             in the list is the number of genomes in the dataset.  Each numpy
+             array contains the string Ensembl ID of genes in the genome, which
+             also correspond to the columns in the out['matrix'].
+        out['feature_types']: List of numpy arrays, where the number of elements
+             in the list is the number of genomes in the dataset.  Each numpy
+             array contains the string feature types of genes (or possibly
+             antibody capture reads), which also correspond to the columns
+             in the out['matrix'].
+
+    """
+    logging.info(f"Detected AnnData format")
+    
+    adata = anndata.read_h5ad(filename)
+    
+    if "counts" in adata.layers.keys():
+        # this is a common manual setting for users of scVI
+        # given the manual convention, we prefer this matrix to 
+        # .X since it is less likely to represent something other
+        # than counts
+        logging.info("Found `.layers['counts']. Using for count data.")
+        count_matrix = adata.layers["counts"]
+    elif "spliced" in adata.layers.keys():
+        # alignment using kallisto | bustools with intronic counts
+        # does not populate `.X` by default, but does populate
+        # `.layers['spliced'], .layers['unspliced']`.
+        # we use spliced counts for analysis
+        logging.info("Found `.layers['spliced']. Using for count data.")
+        count_matrix = adata.layers["spliced"]
+    else:
+        count_matrix = adata.X
+        
+    # check that `count_matrix` contains a large number of barcodes,
+    # consistent with a raw single cell experiment
+    n_expected_barcodes = 1e5
+    if count_matrix.shape[0] < n_expected_barcodes:
+        # this experiment might be prefiltered
+        msg = f"Only {count_matrix.shape[0]} barcodes were found.\n"
+        msg += "This suggests the matrix was prefiltered.\n"
+        msg += "CellBender requires a raw, unfiltered [Barcodes, Genes] matrix."
+        logging.warning(msg)
+        
+    # AnnData is [Cells, Genes], no need to transpose
+    count_matrix = sp.csr_matrix(count_matrix)
+    # feature names and ids are not consistently delineated in AnnData objects
+    # so we report both as the name of column annotations `.var_names`
+    feature_names = np.array(adata.var_names)
+    feature_ids = np.array(adata.var_names)
+    # feature types aren't consistently encoded for AnnData objects
+    feature_types = [np.array(['unknown']*len(adata.var_names))]
+    barcodes = np.array(adata.obs_names)
+    # AnnData does not consistently encode the genome
+    genomes = np.array(['unknown'])
```

Comment:
So this part with feature_types, feature_ids, and genomes is where I get a little concerned.  I know that `anndata` does not specify that you must have these fields, so a lot of people might not.  But what if they are there?  People will probably want some effort made to keep them around.  What about this?

```
feature_names = np.array(adata.var_names)
barcodes = np.array(adata.obs_names)

# Make an attempt to find feature_IDs if they are present.
feature_ids = np.empty(adata.var_names.size, dtype=str)
if 'gene_ids' in adata.var.keys():
    feature_ids = adata.var['gene_ids']
elif 'gene_id' in adata.var.keys():
    feature_ids = adata.var['gene_id']

# Make an attempt to find feature_types if they are present.
feature_types = np.empty(adata.var_names.size, dtype=str)
if 'feature_types' in adata.var.keys():
    feature_types = adata.var['feature_types']
elif 'feature_type' in adata.var.keys():
    feature_types = adata.var['feature_type']

# Make an attempt to find genomes if they are present.
genomes = np.empty(adata.var_names.size, dtype=str)
if 'genomes' in adata.var.keys():
    genomes = adata.var['genomes']
elif 'genome' in adata.var.keys():
    genomes = adata.var['genome']
```

Reply:
This seems perfectly sensible to me. As a future TODO, it might be useful to add additional variables in `.var` to the output dict too. I use a wrapper around `kallisto` that adds on all the attributes from a GTF (e.g. `start`, `chr`, `gene_biotype` et. al.), and I might not be the only one. That might be a safer way to preserve information as the number of ways to align reads grows.